### DISCUSSION TITLE
Speed up random forest with indicator features

### DISCUSSION THREAD

**Discussion by S90pngprgm** [kogalur/randomForestSRC#207] [Q&A]
Given a set of explanatory variables X1, ..., Xn, instead of learning a random forest model to predict the response Y as a function of (X1, ...  Xn), I want to define a new feature X'1 as an indicator variable X'1=Ind(X1>x1), where x1 is constant. For any given x1 value, I could run a random forest for Y using Ind(X1>x1), X2, ..., Xn as features. I want to find the prediction for arbitrary x1. But It would be inefficient to build a random forest for each value of x1.

Is there a way to somehow consider the commonality of this group of random forest on different values of x1 so that they can be generated efficiently? A method is preferred if its time complexity less than linear (ideally independent) of the number of x1 values.

**Comment by ishwaran** [comment]
Sounds to me like you want to use partial prediction implemented by the function `partial.rfsrc`

Consider the following example applied to the `airquality` data set

```
airq.obj <- rfsrc(Ozone ~ ., data = airquality)
     
## partial effect for wind
partial.obj <- partial(airq.obj,
                       partial.xvar = "Wind",
                       partial.values = airq.obj$xvar$Wind)
pdta <- get.partial.plot.data(partial.obj)
     
## plot partial values
plot(pdta$x, pdta$yhat, type = "b", pch = 16,  xlab = "wind", ylab = "partial effect of wind")
```

The plot shows the predicted value for various values of wind, averaged over the other features.  The values of wind we seek predicted values for is specified by the option `partial.values`.  See the [partial.rfsrc](https://luminwin.github.io/randomForestSRC/reference/partial.rfsrc.html)  help file for details.

Suppose you want to work out the mean predicted value for wind larger than say 10.   Then you would just average the partial predicted value (vertical axis value) over those values for wind (horizontal axis) for which wind is larger than 10.

Note also that  `get.partial.plot.data` is a convenient function for parsing `partial.obj` but you can extract the raw predicted values from the latter object to customize what you want to get.  

So in other words, what I'm suggesting is that you can use `partial.rfsrc` which is a fast interface for extracting predicted values.
